core/src/consensus/light_client/mod.rs

```diff
@@ -17,14 +17,50 @@
 pub mod verification;
 pub use self::verification::{verify_header, verify_state, verify_transaction};
 pub use ctypes::BlockNumber;
-pub use primitives::H256;
+pub use primitives::{H256, H512};
+use ckey::SchnorrSignature;
+use rlp::{Decodable, DecoderError, Encodable, Rlp, RlpStream};
+
+#[derive(PartialEq, Debug)]
+pub struct Seal(Vec<(usize, SchnorrSignature)>); // (index_in_vset, sign); schnorr scheme
+
+impl Encodable for Seal {
+    fn rlp_append(&self, s: &mut RlpStream) {
+        s.begin_list(self.0.len() * 2);
+        for (x, y) in self.0.iter() {
+            s.append(x);
+            s.append(&H512::from(*y));
+        }
+    }
+}
+impl Decodable for Seal {
+    fn decode(rlp: &Rlp<'_>) -> Result<Self, DecoderError> {
+        let item_count = rlp.item_count()?;
+        if item_count % 2 == 1 {
+            return Err(DecoderError::RlpInvalidLength {
+                expected: item_count + 1,
+                got: item_count,
+            })
+        }
+        let mut vec = Vec::with_capacity(item_count / 2);
+        for i in 0..(item_count / 2) {
+            let f: H512 = rlp.val_at(i * 2 + 1)?;
+            vec.push((rlp.val_at(i * 2)?, SchnorrSignature::from(f)));
+        }
+        Ok(Self(vec))
+    }
+}
 
-#[derive(RlpEncodable, RlpDecodable, PartialEq, Debug)]
-pub struct Seal(Vec<bool>, H256); // tentative depending on Ed25519 merging
 
 #[derive(RlpEncodable, RlpDecodable, PartialEq, Debug)]
-pub struct LightHeader{
+pub struct CoreHeader{
     pub number: BlockNumber,
     pub hash: H256,
     pub next_validator_set_hash: H256
 }
+
+pub struct LightHeader {
+    pub core: CoreHeader,
+    pub state_root: H256,
+    pub transactions_root: H256,
+}
```

Comment:
How the CoreHeader and LightHeader different?
Why do we need two types of headers?

Reply:
CoreHeader is essential info to update light client state. (ICS uses this) Light header is for tx/state verification, including CoreHeader. (Standalone light client uses this)

Reply:
I think we can have a better name, I'll try to find a good name for the headers.